### PR TITLE
[export] min/max ranges for dim hints

### PR DESCRIPTION
**Pull request by pianpwk**
Differential Revision: D71522032

Adds min/max ranges to Dim.AUTO/DYNAMIC/STATIC, so users can do `Dim.AUTO(min=2, max=2048)`.

### CODE REVIEW

File: torch/export/dynamic_shapes.py [hunk 1]
```diff
@@ -69,6 +72,14 @@ def DYNAMIC():
     def STATIC():
         return _DimHint(_DimHintType.STATIC)
 
+    def __call__(self, min=None, max=None) -> "_DimHint":
+        if not self._factory:
+            raise NotImplementedError
+        assert min is None or min >= 0, "min must be non-negative"
+        assert max is None or max >= 0, "max must be non-negative"
+        assert min is None or max is None or min <= max, "min must be <= max"
+        return _DimHint(self.type, min=min, max=max, _factory=False)
```
Comment (pianpwk) [inline torch/export/dynamic_shapes.py:81]: this is sneaky, but adding this `_factory` field so users can do `Dim.AUTO`, `Dim.AUTO()`, or `Dim.AUTO(min=2, max=2048)` and get the same object class, but raise an error if they do `Dim.AUTO()()`

File: torch/_export/non_strict_utils.py [hunk 2]
```diff
@@ -394,24 +399,53 @@ def make_constraints(
...
+                    if is_int(d):
+                        trace_vr & user_vr
+                    else:
+                        range_constraints[d.node.expr] &= user_vr
+                except torch.utils._sympy.value_ranges.ValueRangeError as exc:
+                    raise ValueError(
+                        f"Received user-specified min/max range of [{dim.min}, {dim.max}], "
+                        f"which conflicted with the inferred min/max range of [{trace_vr.lower}, {trace_vr.upper}] "
+                        f"from model tracing, for inputs{pytree.keystr(flat_paths[input_index])}.shape[{i}]."
+                    ) from exc
```
Comment (angelayi) [inline torch/_export/non_strict_utils.py:447]: I'm a little confused when this error happens. My understanding is that this function is just getting range constraints, but this error seems like a ConstraintViolationError which I would've expected to be thrown earlier?

File: test/export/test_export.py [hunk 3]
```diff
@@ -2371,6 +2371,76 @@ def forward(self, x, y, z):
...
+
+        x, y = torch.randn(8), torch.randn(8)
+
+        # conflict with lower bound
+        shapes = torch.export.ShapesCollection()
+        shapes[x] = [Dim.DYNAMIC(max=5)]
+        with self.assertRaisesRegex(
+            ValueError,
+            r"Received user-specified .* \[None, 5\], which conflicted with the inferred .*"
+            r"\[6, int_oo\] .*, for inputs\['xs'\]\['data'\]\[0\]\[0\]\.shape\[0\]",
```
Comment (ColinPeppler) [inline test/export/test_export.py:2417]: questions from a UX perspective (it's okay if no good answer)
- why is this the error we get?
- would it be possible for the error to mention that max=5, but the x is `torch.Size([8])`?

Comment (pianpwk) [reply test/export/test_export.py:2417]: Here model tracing figured out the valid range is [6, inf], because of the `x[6:]` call, while user specified [0, 5]. These ranges don't overlap, so we raise this error.

We could definitely check the sample input shapes too, but that'd be a separate check (we'd check sample input conflicts first, then check inferred range vs. user range).

I'm curious (totally open question), is there any error message you'd like to see / UX you'd like to have here? I kind of wrote this generically.

File: torch/_export/non_strict_utils.py [hunk 4]
```diff
@@ -359,17 +360,21 @@ def make_constraints(
...
+            isinstance(x, torch.SymInt) and x.node.expr.is_number
+        )
+
     shape_env = fake_mode.shape_env
     assert shape_env is not None
     inline_constraints = gm.meta.get("inline_constraints", [])
-    range_constraints = {
-        symbol: inline_constraints[symbol] for symbol in inline_constraints
-    }
+    range_constraints = defaultdict(lambda: ValueRanges(0, int_oo)) | inline_constraints
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:371]: I wanted this to be a defaultdict haha

File: torch/_export/non_strict_utils.py [hunk 5]
```diff
@@ -394,24 +399,53 @@ def make_constraints(
...
+            if not is_int(d):
                 # Look up the range constraint for the symbol corresponding to this shape dimension
                 # and store it indexed by the symbolic expression corresponding to it.
                 # NOTE(avik): Use node._expr instead of node.expr for the lookup here because
                 # we want the symbol, not its replacement, which could be an expression. Maybe
                 # there's a better way to do this, e.g., by (re)computing value ranges for expressions?
-                dim = shape_spec[i] if shape_spec else None
                 if dim is None or isinstance(dim, _DimHint):
-                    range_constraints[d.node.expr] = shape_env.var_to_range[
+                    range_constraints[d.node.expr] &= shape_env.var_to_range[
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:414]: when we go through the placeholder symbols & dynamic shapes specs, we might see the same symbols over and over, with new min/max ranges, e.g.
```
sizes: [s0, s1], [s0, 4]
shapes: [AUTO(min=4), AUTO], [AUTO(max=32), None]
```

each time we see a symbol (s0), we wanna refine it with both the inferred range (from ShapeEnv), and the user-specified range (dim.min/max). The `&` just takes the intersection of ranges.

File: torch/_export/non_strict_utils.py (hunk 2 above)
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:447]: Yeah, this is a bit weird, but I couldn't find a good way to feed these ranges into produce_guards, so this won't show up for the constraint solver / suggested fixes. We could improve and print out all the violations we hit, and not just the first though.

I did want to keep this separate, since in constraint violations it's a lot of user work to track down where the min/max bound guards are emitted, and here we just know it's from the user-specification.

Is there any error message / UX you'd suggest though? very open

File: torch/_export/non_strict_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/non_strict_utils.py:414]: fancy

File: test/export/test_export.py (hunk 3 above)
Comment (ColinPeppler) [reply test/export/test_export.py:2417]: Thanks for the explanation! I see so I see it's smth like if `[user_specified_min, user_specified_max]` is not a subset of the `[inferred_min, inferred_max]`, then Export throws.

> We could definitely check the sample input shapes too, but that'd be a separate check (we'd check sample input conflicts first, then check inferred range vs. user range).

Yeah, that makes sense. I was expecting whether I'm using Dim Hints or using Named Dim, the type of checks and when the checks happen would be the same.

>  is there any error message you'd like to see / UX you'd like to have here?
I think the current error message is fine! Though, I'm wondering if it's more expedient to throw an error, at the start, mentioning the sample input shapes don't match with the user-specified range.

File: torch/_export/non_strict_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/non_strict_utils.py:414]: is it ever possible that you get one that says AUTO(min=4) and then another one that says AUTO(min=5)?

File: torch/_export/non_strict_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/non_strict_utils.py:371]: more fancy python

File: torch/_export/non_strict_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/non_strict_utils.py:414]: Does & know to take the min of 4 and 5? sorry.. im not familiar with &

[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:414]: ah yep, the final thing would be the intersection of [2, inf] & [4, inf] & [5, inf], so we have a [5, inf] bound

File: torch/_export/non_strict_utils.py [hunk 6]
```diff
@@ -394,24 +399,53 @@ def make_constraints(
...
+                        upper=int_oo if dim.max is None else dim.max,
+                    )
+                    if is_int(d):
+                        trace_vr & user_vr
+                    else:
+                        range_constraints[d.node.expr] &= user_vr
+                except torch.utils._sympy.value_ranges.ValueRangeError as exc:
+                    raise ValueError(
+                        f"Received user-specified min/max range of [{dim.min}, {dim.max}], "
+                        f"which conflicted with the inferred min/max range of [{trace_vr.lower}, {trace_vr.upper}] "
```
Comment (tugsbayasgalan) [inline torch/_export/non_strict_utils.py:445]: What happens when the inferred dim is static? I think this error message is going to be weird since the lower and upper are same numbers?

File: torch/_export/non_strict_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/_export/non_strict_utils.py:371]: This is pretty cool lol

**Review by tugsbayasgalan — APPROVED** [review]

File: torch/_export/non_strict_utils.py (hunk 6 above)
Comment (pianpwk) [reply torch/_export/non_strict_utils.py:445]: Yeah, with AUTO this would report something like, user-specified [8, 12], while inferred [4, 4].

Comment (tugsbayasgalan) [reply torch/_export/non_strict_utils.py:445]: I think we should suggest that it is static in this case, otherwise it is bit confusing
